File: app/src/main/java/com/pandatone/kumiwake/setting/DBBackup.java
package com.pandatone.kumiwake.setting;

import android.content.Context;
import android.os.Environment;
import android.widget.Toast;

import com.pandatone.kumiwake.R;
import com.pandatone.kumiwake.member.GroupListAdapter;
import com.pandatone.kumiwake.adapter.MemberListAdapter;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.nio.channels.FileChannel;

public class DBBackup {
    static String sd_dir, sd_stt, dir_path, mb_db_file, gp_db_file;
    static boolean b;

    public static void DBBackup(Context c) {
        getPath(c);

        File f = new File(dir_path);
        b = f.exists();           //SDカードにkumiwakeディレクトリがあるか。
        if (b == false) {          //ディレクトリが存在しないので作成。
            b = f.mkdirs();    //　sdcard/kumiwakeディレクトリを作ってみる。
            if (b == false) {
                Toast.makeText(c, SettingHelp.getContext().getString(R.string.failed_back_up), Toast.LENGTH_SHORT).show();
                return;         //ディレクトリ作成失敗
            }
        }

        filecopy(mb_db_file, dir_path + "/mb.db", c, true);
        filecopy(gp_db_file, dir_path + "/gp.db", c, true);//DBのファイルをSDにコピー
    }

    public static void DBImport(Context c) {
        getPath(c);

        File f = new File(dir_path);
        b = f.exists();           //SDカードにkumiwakeディレクトリがあるか。
        if (b == false) {
            Toast.makeText(c, SettingHelp.getContext().getString(R.string.not_exist_file), Toast.LENGTH_SHORT).show();
            return;
        }

        filecopy(dir_path + "/mb.db", mb_db_file, c, false);
        filecopy(dir_path + "/gp.db", gp_db_file, c, false);//DBのファイルをインポート
    }


    //ファイルのコピー（チャネルを使用）
    public static void filecopy(String file_src, String file_dist, Context c, Boolean backup) {
        int err;
        FileInputStream fis;
        FileOutputStream fos;

        err = 0;
        File fi = new File(file_src);
        File fo = new File(file_dist);
        try {
            fis = new FileInputStream(fi);
            FileChannel chi = fis.getChannel();

            fos = new FileOutputStream(fo);
            FileChannel cho = fos.getChannel();

            chi.transferTo(0, chi.size(), cho);
            chi.close();
            cho.close();
        } catch (FileNotFoundException e) {
            err = 1;
        } catch (IOException e) {
            err = 2;
        }
        if (err == 0 && backup) {
            Toast.makeText(c, SettingHelp.getContext().getString(R.string.back_up_completed), Toast.LENGTH_SHORT).show();
        } else if (err == 0 && !backup) {
            Toast.makeText(c, SettingHelp.getContext().getString(R.string.import_completed), Toast.LENGTH_SHORT).show();
        }
    }

    public static void getPath(Context c) {
        sd_dir = Environment.getExternalStorageDirectory().getPath();     //SDカードディレクトリ
        sd_stt = Environment.getExternalStorageState();           //SDカードの状態を取得
        MemberListAdapter dbAdapter = new MemberListAdapter(SettingHelp.getContext());
        GroupListAdapter gpdbAdapter = new GroupListAdapter(SettingHelp.getContext());
        dbAdapter.open();
        dbAdapter.getAllNames();
        dbAdapter.close();
        gpdbAdapter.open();
        gpdbAdapter.getAllNames();
        gpdbAdapter.close();
        mb_db_file = MemberListAdapter.db.getPath();
        gp_db_file = GroupListAdapter.db.getPath();   //DBのディレクトリとファイル名
        dir_path = sd_dir + "/KUMIWAKE_Backup";

        b = sd_stt.equals(Environment.MEDIA_MOUNTED);     //SDカードの状態
        if (b == false) {  //書込み状態でマウントされていない。
            Toast.makeText(c, SettingHelp.getContext().getString(R.string.not_mounted), Toast.LENGTH_SHORT).show();
            return;         //ディレクトリ作成失敗
        }

    }

}

File: app/src/main/java/com/pandatone/kumiwake/setting/DBBackup.kt
package com.pandatone.kumiwake.setting

import android.annotation.SuppressLint
import android.content.ContentValues.TAG
import android.content.Context
import android.util.Log
import com.pandatone.kumiwake.adapter.GroupAdapter
import com.pandatone.kumiwake.adapter.MemberAdapter
import com.pandatone.kumiwake.history.HistoryAdapter
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException
import java.util.zip.ZipEntry
import java.util.zip.ZipInputStream
import java.util.zip.ZipOutputStream

@SuppressLint("StaticFieldLeak")
object DBBackup {
    private val mb_db_file: String
        get() {
            MemberAdapter(context).also {
                it.open()
                it.getDB
                it.close()
            }
            return MemberAdapter.db.path
        }
    private val gp_db_file: String
        get() {
            GroupAdapter(context).also {
                it.open()
                it.getDB
                it.close()
            }
            return GroupAdapter.db.path   //DBのディレクトリとファイル名
        }
    private val hs_db_file: String
        get() {
            HistoryAdapter(context).also {
                it.open()
                it.getDB
                it.close()
            }
            return HistoryAdapter.db.path   //DBのディレクトリとファイル名
        }
    private lateinit var context: Context

    private fun getFileNameByDB(dbFilePath: String): String {
        return when (dbFilePath) {
            File(mb_db_file).name -> "mb.db"
            File(gp_db_file).name -> "gp.db"
            File(hs_db_file).name -> "hs.db"
            else -> {
                throw Exception("Invalid File")
            }
        }
    }

    private fun getDBByFileName(fileName: String): String {
        return when (fileName) {
            "mb.db" -> mb_db_file
            "gp.db" -> gp_db_file
            "hs.db" -> hs_db_file
            else -> {
                throw Exception("Invalid File Name")
            }
        }
    }

    // Backup
    fun addBackupFiles(c: Context, zipOutputStream: ZipOutputStream) {
        context = c
        val backupFile = createFolderWithFiles(context, mb_db_file, gp_db_file, hs_db_file)
        addFolderToZip(backupFile, zipOutputStream)
        zipOutputStream.finish()
    }

    private fun createFolderWithFiles(context: Context, vararg filePaths: String): File {
        val folderName = "kumiwake_backup"
        val filesDir = context.getExternalFilesDir(null)
        val folder = File(filesDir, folderName)

        if (!folder.exists()) {
            val created = folder.mkdirs()
            if (!created) {
                Log.e(TAG, "Cannot create backup folder")
            }
        }

        for (filePath in filePaths) {
            val file = File(filePath)
            if (file.exists()) {
                val destinationFile = File(folder, getFileNameByDB(file.name))
                try {
                    file.copyTo(destinationFile, overwrite = true)
                } catch (e: IOException) {
                    Log.e(TAG, "Failed to copy file: ${file.name}", e)
                }
            }
        }

        return folder
    }

    private fun addFolderToZip(folder: File, zipOutputStream: ZipOutputStream) {
        folder.listFiles()?.forEach { file ->
            val filePath = file.name
            try {
                FileInputStream(file).use { fileInputStream ->
                    val zipEntry = ZipEntry(filePath)
                    zipOutputStream.putNextEntry(zipEntry)
                    fileInputStream.copyTo(zipOutputStream)
                    zipOutputStream.closeEntry()
                }
            } catch (e: IOException) {
                Log.e(TAG, "Failed to add file to zip: $filePath", e)
            }
        }
    }

    // Import
    fun readFile(c: Context, zipInputStream: ZipInputStream) {
        context = c
        val buffer = ByteArray(1024)
        var entry = zipInputStream.nextEntry
        while (entry != null) {
            val fileName = entry.name
            val file = File(getDBByFileName(fileName))
            val outputStream = FileOutputStream(file)
            var len: Int
            while (zipInputStream.read(buffer).also { len = it } > 0) {
                outputStream.write(buffer, 0, len)
            }
            outputStream.close()
            entry = zipInputStream.nextEntry
        }
        zipInputStream.closeEntry()
        zipInputStream.close()
    }

}
